File: 34-practical-oojs/oo/dog.js
class Dog {

  // class variable
  static all = []

  constructor(firstName, favSnacks) {
    this.firstName = firstName
    this.favSnacks = favSnacks
    Dog.all.push(this)
  }

  // instance method
  speak() {
    console.log(this)
    console.log(`Woof my name is ${this.firstName}`)
  }

  // class method
  static findByName(name) {
    return Dog.all.find(dogInstance => {
      return dogInstance.firstName === name
    })
  }
}

// Dog.all = []

// @@all
// class methods
// getters/setters

const fezzik = new Dog(
  "Fezzik", 
  ["Peanut Butter", "Soccer Balls"]
)
const lucy = new Dog(
  "Lucy", 
  ["Dog Treats", "Turkey Sandwiches"]
)

fezzik.speak()

const speak = fezzik.speak

File: 34-practical-oojs/oo/dog.rb
class Dog
  attr_accessor :first_name, :fav_snacks

  @@all = []

  def initialize(first_name, fav_snacks)
    @first_name = first_name
    @fav_snacks = fav_snacks
    @@all << self
  end

  def speak
    puts "Woof my name is #{self.first_name}"
  end

  def self.all
    @@all
  end

  def self.find_by_name(name)
    self.all.find do |dog_instance|
      dog_instance.first_name == name
    end
  end

end

fezzik = Dog.new(
  "Fezzik", 
  ["Peanut Butter", "Soccer Balls"]
)
fezzik.speak

lucy = Dog.new(
  "Lucy", 
  ["Dog Treats", "Turkey Sandwiches"]
)


Dog.find_by_name("Lucy")

speak = fezzik.speak